astrapy/api.py

```diff
@@ -0,0 +1,91 @@
+import logging
+import httpx
+from typing import Any, Dict, Optional, TypeVar, cast
+
+from astrapy.types import API_RESPONSE
+from astrapy.utils import make_request
+
+T = TypeVar("T", bound="APIRequestHandler")
+
+
+logger = logging.getLogger(__name__)
+
+
+class APIRequestError(ValueError):
+    def __init__(self, response: httpx.Response) -> None:
+        self.response = response
+        super().__init__("The API produced an error.")
+
+    def __repr__(self) -> str:
+        return f"{self.response}"
+
+    def get(self) -> API_RESPONSE:
+        return {
+            "error": True,
+            "status_code": self.response.status_code,
+            "response_body": self.response.json(),
+            "response": self.response,
+        }
+
+
+class APIRequestHandler:
+    def __init__(
+        self: T,
+        client: httpx.Client,
+        base_url: str,
+        auth_header: str,
+        token: str,
+        method: str,
+        json_data: Optional[Dict[str, Any]],
+        url_params: Optional[Dict[str, Any]],
+        path: Optional[str] = None,
+        skip_error_check: bool = False,
+    ) -> None:
+        self.client = client
+        self.base_url = base_url
+        self.auth_header = auth_header
+        self.token = token
+        self.method = method
+        self.path = path
+        self.json_data = json_data
+        self.url_params = url_params
+        self.skip_error_check = skip_error_check
+
+    def _raw_request(self: T) -> httpx.Response:
+        return make_request(
+            client=self.client,
+            base_url=self.base_url,
+            auth_header=self.auth_header,
+            token=self.token,
+            method=self.method,
+            path=self.path,
+            json_data=self.json_data,
+            url_params=self.url_params,
+        )
+
+    def request(self: T) -> API_RESPONSE:
+        # Make the raw request to the API
+        self.response = self._raw_request()
+
+        # If the response was not successful (non-success error code) raise an error directly
+        self.response.raise_for_status()
+
+        # Otherwise, process the successful response
+        return self._process_response()
+
+    def _process_response(self: T) -> API_RESPONSE:
+        # In case of other successful responses, parse the JSON body.
+        try:
+            # Cast the response to the expected type.
+            response_body: API_RESPONSE = cast(API_RESPONSE, self.response.json())
+
+            # If the API produced an error, warn and return the API request error class
+            if "errors" in response_body and not self.skip_error_check:
```

Comment:
as discussed elsewhere, not all the error responses contain an "errors" key,
probably we need to handle pass a function "response_contains_errors" and implement custom handling for each different API call

ideally the API should return a non 200 HTTP Code in case of errors. I know that this is not the case, so here in AstraPy we cannot do much

Maybe now that we are in "preview" we can raise the problem in Stargate and fix it upstream (cc @clun @prabhatjha )

Reply:
Updated the logger, but yes.... i'm not sure exactly all the possible error conditions that can occur and how we know they're errors exactly. It feels like if possible this would be ideal to fix upstream for sure, or at least if its well documented somewhere and a good reason for what possible errors we might get and the format of said responses.